internal/operator/statefulset/statefulset.go

```diff
@@ -0,0 +1,216 @@
+	}
+}
+
+// GetExistingResourceNames returns the name of the existing statefulset for the given Etcd.
+func (r _resource) GetExistingResourceNames(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) ([]string, error) {
+	resourceNames := make([]string, 0, 1)
+	objectKey := getObjectKey(etcd)
+	sts, err := r.getExistingStatefulSet(ctx, etcd)
+	if err != nil {
+		return nil, druiderr.WrapError(err,
+			ErrGetStatefulSet,
+			"GetExistingResourceNames",
+			fmt.Sprintf("Error getting StatefulSet: %v for etcd: %v", objectKey, etcd.GetNamespaceName()))
+	}
+	if sts == nil {
+		return resourceNames, nil
+	}
+	if metav1.IsControlledBy(sts, etcd) {
+		resourceNames = append(resourceNames, sts.Name)
+	}
+	return resourceNames, nil
+}
+
+// Sync creates or updates the statefulset for the given Etcd.
+func (r _resource) Sync(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) error {
+	var (
+		existingSTS *appsv1.StatefulSet
+		err         error
+	)
+	objectKey := getObjectKey(etcd)
+	if existingSTS, err = r.getExistingStatefulSet(ctx, etcd); err != nil {
+		return druiderr.WrapError(err,
+			ErrSyncStatefulSet,
+			"Sync",
+			fmt.Sprintf("Error getting StatefulSet: %v for etcd: %v", objectKey, etcd.GetNamespaceName()))
+	}
+	// There is no StatefulSet present. Create one.
+	if existingSTS == nil {
+		return r.createOrPatch(ctx, etcd)
+	}
+
+	// StatefulSet exists, check if TLS has been enabled for peer communication, if yes then it is currently a multistep
+	// process to ensure that all members are updated and establish peer TLS communication.
+	if err = r.handlePeerTLSChanges(ctx, etcd, existingSTS); err != nil {
+		return druiderr.WrapError(err,
+			ErrSyncStatefulSet,
+			"Sync",
+			fmt.Sprintf("Error while handling peer URL TLS change for StatefulSet: %v, etcd: %v", objectKey, etcd.GetNamespaceName()))
+	}
+	return r.createOrPatch(ctx, etcd)
+}
+
+// TriggerDelete triggers the deletion of the statefulset for the given Etcd.
+func (r _resource) TriggerDelete(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) error {
+	objectKey := getObjectKey(etcd)
+	ctx.Logger.Info("Triggering delete of StatefulSet", "objectKey", objectKey)
+	err := r.client.Delete(ctx, emptyStatefulSet(etcd))
+	if err != nil {
+		if errors.IsNotFound(err) {
+			ctx.Logger.Info("No StatefulSet found, Deletion is a No-Op", "objectKey", objectKey.Name)
+			return nil
+		}
+		return druiderr.WrapError(err,
+			ErrDeleteStatefulSet,
+			"TriggerDelete",
+			fmt.Sprintf("Failed to delete StatefulSet: %v for etcd %v", objectKey, etcd.GetNamespaceName()))
+	}
+	ctx.Logger.Info("deleted", "component", "statefulset", "objectKey", objectKey)
+	return nil
+}
+
+func (r _resource) getExistingStatefulSet(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) (*appsv1.StatefulSet, error) {
+	sts := emptyStatefulSet(etcd)
+	if err := r.client.Get(ctx, getObjectKey(etcd), sts); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
```

Comment:
no, that is not correct. If it is not found then no error should be returned.

Reply:
The intent of keeping the `errors.IsNotFound(err)` check within the `getExistingStatefulSet()` method is to not leak implementation details outside of the method. So although this forces the caller to check for a nil error as well as a nil sts before making any decision, it's intended behavior. I could add a docstring for the method to indicate this behavior, so that callers don't make mistakes in handling the error and return value for the sts.